core/src/main/java/org/wildfly/plugin/core/DeploymentManager.java

```diff
@@ -0,0 +1,203 @@
+     *
+     * @param deployments a set of deployments to redeploy
+     *
+     * @return the result of the deployment
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    DeploymentResult redeploy(Set<Deployment> deployments) throws IOException;
+
+    /**
+     * Undeploys the deployment from the server.
+     *
+     * @param undeployDescription the description for undeploying the content
+     *
+     * @return the result of the deployment
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    DeploymentResult undeploy(UndeployDescription undeployDescription) throws IOException;
+
+    /**
+     * Undeploys the deployment from the server.
+     *
+     * @param undeployDescriptions the descriptions for undeploying the content
+     *
+     * @return the result of the deployment
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    DeploymentResult undeploy(Set<UndeployDescription> undeployDescriptions) throws IOException;
+
+    /**
+     * Returns the available deployments.
+     *
+     * @return the deployments
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    Set<DeploymentDescription> getDeployments() throws IOException;
+
+    /**
+     * Returns all the deployments on the specified server-group. These deployments may also belong to other server
+     * groups.
+     *
+     * @param serverGroup the server group to get the deployments for
+     *
+     * @return the deployments
+     *
+     * @throws IOException           if a failure occurs communicating with the server
+     * @throws IllegalStateException if the running server is not a managed domain
+     */
+    Set<DeploymentDescription> getDeployments(String serverGroup) throws IOException;
+
+    /**
+     * Returns the names of deployed content.
+     *
+     * @return the names of deployed content
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    Set<String> getDeploymentNames() throws IOException;
+
+    /**
+     * Checks if the deployment is on the server.
+     *
+     * @param name the name of the deployment
+     *
+     * @return {@code true} if the deployment exists otherwise {@code false}
+     *
+     * @throws IOException if a failure occurs communicating with the server
+     */
+    boolean hasDeployment(String name) throws IOException;
+
+    /**
+     * Checks if the deployment is on the server.
+     *
+     * @param name        the name of the deployment
+     * @param serverGroup the server group to check for the deployment on
+     *
+     * @return {@code true} if the deployment exists otherwise {@code false}
+     */
+    boolean hasDeployment(String name, String serverGroup) throws IOException;
+
+    /**
+     * A factory to create a new deployment manager
+     */
+    class Factory {
```

Comment:
You may well be right in this particular situation (that I have not looked at in detail). The question about dependency injection just came to my mind, when I saw the Factory - because Factories are generally often eliminated by using DI. I must say I am not sure myself, if DI is a good idea for a lib like this that is supposed to be used in Maven, Gradle, SBT and who knows where else, where a DI container is not always available. 

And finally, to make the picture complete, why I am bringing this topic: Since longer, I planned the very same refactoring in my srcdeps-maven-plugin[https://github.com/l2x6/srcdeps-maven-plugin] . Much like you, I am introducing a srcdeps-core lib so that it can be used not only from Maven but also from Gradle and other tools. 
I started two days ago, and I have chosen to use the DI. I am curious to figure out myself if that turns out to be a good idea :) Here is the working branch if you happen to want to have a look: https://github.com/ppalaga/srcdeps-maven-plugin/tree/160613-introduce-core